### PR TITLE
invalid state.save() results, option  1.

### PR DESCRIPTION
**Pull request by rerdavies**
Addresses issue #36.

- state restore loads the model synchronously if not activated.
- State restore should clear the model if a model path isn't
  provided in state.
-  Avoid error messages in ardour when loading empty
  state.
- load errors clear the filename, instead of leaving the
  previous filename, while not having a model.

Reflects discussions with @falkTX  . Improved handling of "" in state save/restore, reflecting bugs in Ardour and PiPedal.

Clearing the filename on error has not been discussed, but it seems  preferable to leaving the old filename, but having no model.

### CODE REVIEW

**Comment by rerdavies** [comment]
Fixes #36

File: src/nam_plugin.cpp [hunk 1]
```diff
@@ -87,40 +124,12 @@ namespace NAM {
...
-							model->finalize_(numSamples);
-
-							delete[] buffer;
-						}
-					}
+					::DSP* model = nam->load_model(msg->path);
 
 					LV2SwitchModelMsg response = { kWorkTypeSwitch, {}, model };
-					memcpy(response.path, msg->path, pathlen);
+					memcpy(response.path, msg->path, strlen(msg->path));
```
Comment (falkTX) [inline src/nam_plugin.cpp:132]: this breaks the case of `pathlen == 0 || pathlen >= MAX_FILE_NAME`.
so the `pathlen` should be kept outside of the function, maybe have the model load function only deal with valid paths.

**Comment by falkTX** [comment]
added a little comment for something that needs attention.

your code style doesn't match the current one though, many times you miss a space between the variable type and name, like `::DSP*load_model`.

speaking of space, please cleanup the use of whitespace which changed from the original repo, it makes reviewing any code much harder than it needs to be.
plus reformatting of the code shouldn't be on this PR, leave it for another one or at least separate it with different commits please.

the changes should be as clean as possible to reduce friction and making reviewing easy.
it is not my repo of course, but I wouldn't merge such changes as they are very unclean at the moment.

**Comment by mikeoliphant** [comment]
I'll admit that I haven't looked at this PR very closely for two reasons. First, because of the uncertainty and disagreement over the "correct" way to do things. And second, because of the issues @falkTX raises - PRs are most easy to handle when they are focused on a single issue and do not mess around with the existing formatting.

**Comment by rerdavies** [comment]
@falkTx:
@mikeoliphant :

The whitespace changes are (as far as I am aware) entirely necessary. They reflect indentation changes due to code changes or necessary restructuring of if statements.  The diffs are admittedly rather unpleasant because code had to be refactored out of Plugin::work into a common function, in order to allow patch.restore() to load the model synchronously.  The code is substantive enough that a cut-and-paste would not be appropriate. 

I have corrected instances where there was a missing space after a *.  I am unable to use automatic code reformatting since VSCode's formatting conventions don't match the ones being used. (it puts spaces before the * instead of afterward. If you see any others, feel free to correct them. If you have an automated code formatter that reflects conventions being used in the project, I would be entirely grateful if you could apply it and make any further changes required.

If you have specific suggestions as to how the whitespace changes could be improved, I am open to suggestions. But I do believe that all whitespace changes are neccessary.

I have the style-changes committed on my forked branch. I'm not sure what the easiest way to submit them is.  Do you want me to resubmit the pull-request, or is there a better way to do it?

**Comment by rerdavies** [comment]
Nevermind. The pull request automatically picked up the commit for style changes, so there's nothing extra that needs to be done.

**Comment by mikeoliphant** [comment]
@rerdavies Please remove the version bump.

Beyond that, I will need to rely on agreement between you and @falkTX that these changes are required and the best way to proceed.

**Comment by rerdavies** [comment]
@mikeoliphant : I think you're going to have to decide for yourself at this point. All concerns raised by falcTx have been addressed. I could re-implement Neural Amp Modeler from scratch with less effort than it has taken to get this quite straightforward pull request accepted.

**Comment by rerdavies** [comment]
By way of explanation, the GPL3 license was probably the biggest problem for me. It seems ungrateful to me to take a large body of work with an MIT license and infect it with a GPL 3 license. TooB Neural Amp Modeler allows me to bundle the plugin based on the same compute backend you're using without  fear of infection by GPL3. 

But it certainly didn't help that it TRULY took less time to re-implement from scratch using the MIT-licensed NeuralAmpModelerCore than it did to get this pull request accepted. Less than a day, not including customizations for PiPedal. Although I will admit that I had the advantage of having a robust LV2 plugin base class that you probably didn't.

I do think you need to accept the pull request.  It addresses a bunch of load/save issues, most probably fixes the reported issue with Ardour, and has a good chance of fixing the crash with Reaper.

**Comment by mikeoliphant** [comment]
@rerdavies I understand your frustration, but in the end I'm the one that will be responsible for the code. I have not yet merged this PR for two reasons: 1) the ongoing disagreements between you and @falkTX, and 2) because it does not resolve any issues I have seen with hosts I have tested with (Reaper, Ardour).

I'm not going to get into an argument about licenses. You are free to license your code as you wish, as am I. I think the MIT license is a good one for the NeuralAmpModelerCore and (as you'll see if you look at the GitHub contributors there) I've been happy to contribute under that license.

**Comment by rerdavies** [comment]
@mikeoliphant 

Of the four problems this pull request fixes, three of them occur in Ardour and PiPedal. And the fourth only occurs in PiPedal.

-     Fix: State restore should clear the model if a model path isn't
      provided in state.
  
     -    Load NAM. Don't select a model.
     -    Save a plugin preset ("No Model").
     -   Select a model
     -   Load the "No Model" preset.

     Result: the model selected in the 3rd step remains selected.
     
     The fix: the model should be set to NULL AND an empty patch_set reply should be sent to clear 
     the filname in the UI.

       
- Fix:  Avoid error messages in ardour when loading empty
state. 

    -   Load a plugin preset with no model ("No model") from above.
    -   The plugin logs an error message, which Ardour displays.
    
    Since the LV2 spec says that logged LV2 errors _should_ (must?) generate a user-facing error dialog (which PiPedal does), the 
   user experience is jarring. In the case of Ardour, an error is sent to the log window,
   causing the red LED to flash, which is also jarring. PiPedal, on the other hand, currently suppresses errors that are 
   logged during state.restore(), but I would like to change that behavior. 

   The fix: it is in fact not an error if the state doesn't contain a model filename. Don't log an error.  Or if you must log 
   something, log a warning (which doesn't have to be presented to the user).

-    Fix: load errors clear the filename, instead of leaving the
     previous filename, while not having a model.
     
     - Supply an invalid file to the Model load dialog (e.g. a file that isn't .nam)
     Result: 

      Result: the file control displays the old model, but the model is in fact NULL. The UI state does not reflect the state of the 
      plugin. Saving and loading a plugin preset in this state will result in empty state. 

      The fix: send an empty filename to the UI after a failed state.restore().

The fourth fix: the  result of state.save() is invalid between the initial load of the plugin, and the time that the scheduler task completes. This is an issue for PiPedal specifically, and requires extremely fragile handling for PiPedal and any other hosts that choose to implement a "Patch modifed" indicator.  

This issue does NOT occur with the LV2 sample plugin. There was a one-line fix (option #2), but falkTx chose option #1. 

By that point, the discussion had digressed to questions of what the sample plugin does or does not do. Loading the state synchronously before Activate() is the more complex of the two options, but it is exactly what the reference plugin does. In the interest of expediting the pull request, I deferred. The current pull request does exactly what the reference plugin does. It is not a complicated fix.

Give that falkTx is a principle of MODEP, and I am a principle of PiPedal, which directly competes with MODEP, I have to question whether there is deliberate obstruction going on here. This is a ridiculously simple fix, even with option #1.

**Comment by mikeoliphant** [comment]
> Give that falkTx is a principle of MODEP, and I am a principle of PiPedal, which directly competes with MODEP, I have to question whether there is deliberate obstruction going on here.

To be honest, I've found much of the interaction both ways to be unnecessarily unpleasant, but I believe it is the result of genuine disagreement caused by the vagaries of LV2 and differences in host implementations.

My hope was that, since you have both implemented hosts and I have not, you could agree on the best path forward. That seems not to be the case.

Thank you for the clearer description of the issues you are trying to address. I will look into reproducing them and decide how to proceed from there.

**Comment by rerdavies** [comment]
Ironically, I think we have agreed on a path forward.

There was an earlier comment along the lines of "further comments coming" that was made about 4 days ago. I can't find it anymore. Perhaps it was edited or deleted.

**Comment by falkTX** [comment]
sorry if I need to do other things besides this. you tend to do very long comments which is actually quite detrimental on the motivation side to come back to this. so much time wasted for things that should be quite simple, there is no need to be as verbose as you are with several pages of writing.

anyhow, the first thing to check is the load-directly-if-in-deactivated-state. your PR kinda get us there, but it also adds a couple more extras that need more discussion IMO.

and yes I know what I said, maybe we just have different views on what "later" means. sorry I just dont find this whole interaction very pleasent, and have other things in higher priority. will try to still give a proper reply this week.

**Comment by mikeoliphant** [comment]
I decided to go with the method of updating the currentModelPath in restore(), as the "direct load when not active" approach didn't fix all issues in Reaper (switching presets when the plugin was active didn't update the UI properly).

Commit is here: https://github.com/mikeoliphant/neural-amp-modeler-lv2/commit/33a0c083278c2abed3a4ce202597d6ea4d5a8d7f

From the earlier discussions, I think this also resolved the pi-pedal issues? If not, let me know.

**Comment by rerdavies** [comment]
I reviewed your commit very carefully. It seems to address all four issues. I'm going to close this pull request.  

Thank you!

**Pull request closed without merging**